sdk: Fix decision log masking of input object

epaulson10 opened the pull request:
The sdk package allows you to pass arbitrary Go objects as the input document. This is fine for evaluation (as long as the object can be parsed to an AST), but is problematic for decision log masking, as the masking logic expects the input on the event to be either a map[string]interface{} or a []interface{}. Currently, the decision log masking silently fails if the input object is not of the correct type.

This commit addresses this by checking if the input object is of the correct type, and if not, rebuilds it from the input AST before it is logged.

Also, a test is added to cover this case.

<!--

Thanks for submitting a PR to OPA!

Before pressing 'Create pull request' please read the checklist below.

* All code changes should be accompanied with tests. If you are not
modifying any tests, just provide a short explanation of why updates
to tests are not necessary. In addition to helping catch bugs, tests
are extremely helpful in providing _context_ that explains how your
changes can be used.

* All changes to public APIs **must** be accompanied with
docs. Examples of public APIs include built-in functions,
config fields, and of course, exported Go types/functions/constants/etc.

* Commit messages should explain _why_ you made the changes, not what
you changed. Use active voice. Keep the subject line under 50
characters or so.

* All commits must be signed off by the author. If you are not
familiar with signing off, see our contributor guide below.

For more information on contributing to OPA see:

* [Contributing Guide](https://www.openpolicyagent.org/docs/latest/contributing/)
  for high-level contributing guidelines and development setup.

-->

### Why the changes in this PR are needed?

The SDK package allows you to specify Go structs as the input to a query. However, the decision log masking feature will fail silently when you do so, because the masking only allows the input to be of type `map[string]interface{}` or `[]interface{}`.

<!--
Include a short description of WHY the changes were made.
-->

### What are the changes in this PR?

Adds a test that reproduces this issue.

Checks the type of the provided input object, and if it is not compatible with the decision log masking, it rebuilds the object from the parsed AST of the original input.

<!--
Include a short description of WHAT changes were made.
-->

### Notes to assist PR review:

N/A
<!--
Here you can add information you think will help the reviewer(s).
-->

### Further comments:

The idea to rebuild the object from the AST came from how the HTTP server handles the input.

<!--
Here you can include links to additional resources related to the changes, discuss your solution, other approaches you considered etc.
-->

srenatus commented on sdk/opa.go:
```diff
@@ -344,6 +344,19 @@ func (opa *OPA) executeTransaction(ctx context.Context, record *server.Info, wor
+	// or []interface{}. If the input was not one of those types, rebuild it from the AST.
+	switch (*record.Input).(type) {
+	case map[string]interface{}, []interface{}:
+		break
```
[nit] This `break` statement is superfluous. If it just read
```go
	switch (*record.Input).(type) {
	case map[string]interface{}, []interface{}:
	default:
		// ...
	}
```
it would have the same effect.

srenatus replied:
What about other native types, though? Do we need to worry about, say, `[]string`, which won't be covered by `[]interface{}` here? (Or does that _need_ the same treatment because the decision logger also can't deal with `[]string`? I can't remember 😅)

epaulson10 replied:
Thanks, always having a break is something from C that I have trouble giving up.

> What about other native types, though?

The masking logic only lets you mask keys in objects. You cannot mask elements in slices (although you can reference an object key that is contained in a slice).

To your point on handling native types, I realized that checking for the type of object with the switch was not good enough, as an inner object of the `map[string]interface{}` could` be a map[string]string, for example. So I dropped the switch and am unconditionally doing the conversion. It would be nice to know if masking was enabled so we could avoid this conversion unless it is necessary, but it seems that requires a rego query to determine.

srenatus commented on sdk/opa.go:
```diff
@@ -344,6 +344,19 @@ func (opa *OPA) executeTransaction(ctx context.Context, record *server.Info, wor
+	case map[string]interface{}, []interface{}:
+		break
+	default:
+		asJSON, err := ast.JSON(record.InputAST)
```
💭 I was wondering a bit about some unnecessary work here -- since the execute path needs to do something similar -- but that has already been taken into account: we're using `record.InputAST` here, not `record.Input`. :+1:

epaulson10 replied:
I realized another way to avoid unnecessary work is to move this logic into the if statement that sees if the logger is registered with the manager. There is no sense in doing this unless we are going to actually log something.

srenatus commented on sdk/opa_test.go:
```diff
@@ -1341,6 +1341,84 @@ main = time.now_ns()
+
+	defer opa.Stop(ctx)
+
+	// Verify that timestamp matches time.now_ns() value.
```
[nit] Comment doesn't seem acurate anymore -- also, we don't need to deal with time here, do we?

epaulson10 replied:
You have caught me cargo-culting from another test :see_no_evil:  Fixed.

srenatus reviewed (commented):
This looks good, just questions and nitpicks. Nice catch! And thanks for contributing. 👏

epaulson10 commented:
I amended my commit based on the above discussion. Main points are:

1. I dropped the switch altogether, as checking the "outer type" is not good enough. Through testing I confirmed that any inner maps also need to be `map[string]interface{}`.
2. I moved the conversion from the AST inside the if statement that is checking if the decision log plugin is registered, to avoid doing this conversion unless we actually will be decision logging.
3. Added more to the test.

srenatus reviewed (approved):
Thanks a lot! LGTM
